pkg/R/RDD.R

```diff
@@ -1039,6 +1039,60 @@ setMethod("keyBy",
             lapply(rdd, apply.func)
           })
 
+#' Return a new RDD that has exactly numPartitions partitions.
+#' Can increase or decrease the level of parallelism in this RDD. Internally,
+#' this uses a shuffle to redistribute data.
+#' If you are decreasing the number of partitions in this RDD, consider using
+#' coalesce, which can avoid performing a shuffle.
+#'
+#' @param rdd The RDD.
+#' @param numPartitions Number of partitions to create.
+#' @rdname repartition
+#' @seealso coalesce
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd <- parallelize(sc, list(1, 2, 3, 4, 5, 6, 7), 4L)
+#' numPartitions(rdd)                   # 4
+#' numPartitions(repartition(rdd, 2L))  # 2
+#'}
+setGeneric("repartition", function(rdd, numPartitions) { standardGeneric("repartition") })
+
+#' @rdname repartition
+#' @aliases repartition,RDD
+setMethod("repartition",
+          signature(rdd = "RDD", numPartitions = "integer"),
+          function(rdd, numPartitions) {
+            jrdd <- callJMethod(getJRDD(rdd), "repartition", numPartitions)
+            RDD(jrdd)
+          })
+
+#' Return a new RDD that is reduced into numPartitions partitions.
+#'
+#' @param rdd The RDD.
+#' @param numPartitions Number of partitions to create.
+#' @rdname coalesce
+#' @seealso repartition
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' rdd <- parallelize(sc, list(1, 2, 3, 4, 5), 3L)
+#' numPartitions(rdd)               # 3
+#' numPartitions(coalesce(rdd, 1L)) # 1
+#'}
+setGeneric("coalesce", function(rdd, numPartitions, ...) { standardGeneric("coalesce") })
+
+#' @rdname coalesce
+#' @aliases coalesce,RDD
+setMethod("coalesce",
+           signature(rdd = "RDD", numPartitions = "integer"),
```

Comment:
Actually I agree with @hlin09 that having an integer argument is not ideal. This is mainly because when users try to use the function with say `2` as the second argument, they get an error which is hard to understand because of the way R's dispatch works (See an example for collectPartition below).

```
unable to find an inherited method for function ‘collectPartition’ for signature ‘"RDD", "numeric"’
```

I think the right thing to do is to add a check to see if the argument is not integer and warn the user. Something like
```
if (as.integer(numPartitions) != numPartitions) {
  warning("numPartitions should be an integer. Coercing to integer")
}
```